### PR TITLE
Implementing Equals and GetHashCode for KindAndKey

### PR DESCRIPTION
**Pull request by JeffAshton**
**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission guidelines](../blob/master/CONTRIBUTING.md#submitting-pull-requests)
- [x] I have validated my changes against all supported platform versions

**Related issues**

Performance

**Describe the solution you've provided**

Implemented Equals & GetHasCode instead of relying on auto generated struct implementation.

### CODE REVIEW

File: src/LaunchDarkly.ServerSdk/Internal/DataModelDependencies.cs [hunk 1]
```diff
@@ -16,6 +17,16 @@ internal KindAndKey(DataKind kind, string key)
             Kind = kind;
             Key = key;
         }
+
+        public bool Equals(KindAndKey other)
```
Comment (JeffAshton) [inline src/LaunchDarkly.ServerSdk/Internal/DataModelDependencies.cs:21]: Taken from the identical `struct CacheKey`:

https://github.com/launchdarkly/dotnet-server-sdk/blob/08263f3f1cb2da5dac13075be105515ac9a35d30/src/LaunchDarkly.ServerSdk/Internal/DataStores/PersistentStoreWrapper.cs#L450

File: src/LaunchDarkly.ServerSdk/Internal/DataModelDependencies.cs [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
+using System.Collections.Generic;
 using System.Linq;
 using LaunchDarkly.Sdk.Server.Internal.Model;
 
 using static LaunchDarkly.Sdk.Server.Interfaces.DataStoreTypes;
 
 namespace LaunchDarkly.Sdk.Server.Internal.DataSources
 {
-    internal struct KindAndKey
+    internal readonly struct KindAndKey : IEquatable<KindAndKey>
```
Comment (JeffAshton) [inline src/LaunchDarkly.ServerSdk/Internal/DataModelDependencies.cs:10]: Used as a dictionary and hashset key.  Ran a comparison with 10,000 hashes:

10,000 hashes

Without Fix:
```
5182 ms
5173 ms
5270 ms
5261 ms
5244 ms
5287 ms
5177 ms
5350 ms
5190 ms
5163 ms
```

With Fix:
```
3 ms
0 ms
0 ms
0 ms
1 ms
0 ms
0 ms
0 ms
0 ms
1 ms
```

Originally I tried with 1,000,000 million hashes but got tired of waiting for the old implementation to finish.

Comment (JeffAshton) [reply src/LaunchDarkly.ServerSdk/Internal/DataModelDependencies.cs:10]: What this looks like in production right now when we change a bunch of features together:

image

**Comment by JeffAshton** [comment]
@eli-darkly,   Can I ask you to cut a release today with this performance fix?  We are currently having to scale up servers to handle it.

**Comment by louis-launchdarkly** [comment]
Hello @JeffAshton, thank you for your contribution and also for providing performance metrics to explain your change. While we have started the discussion and are looking at your contribution internally, getting a release out is not a short process that we can always do without some heads-up time, so I can't say we can release today. We would be appreciated it if you notice an issue ahead of time, especially an issue that impacts your server performance, please let us know before you have the PR ready, so we are more prepared to review your contribution and get a release ready.

On a different note, if you don't mind, can you provide your environment information through LaunchDarkly Support to us so we can investigate the performance issue? We think you may have a lot of flags that compound the performance issue, and we may have more suggestions on how you and your company can improve your LaunchDarkly usage experience

**Comment by eli-darkly** [comment]
@JeffAshton The change makes sense and it'll be in the next patch release. I've made it from our end rather than merging this PR, just in order to implement the tests a bit differently, but we'll still acknowledge your PR as usual in the release notes.

**Comment by JeffAshton** [comment]
Thanks @eli-darkly for following up.  I'll fork in the short term.  @louis-launchdarkly's response was pretty infuriating.  The open source allows me to help identity these sorts of problems and share best practices that ultimately help Launch Darkly and its users.

**Comment by eli-darkly** [comment]
@JeffAshton, my opinion is the same as Louis's. No one said you should not submit PRs; we've always appreciated your contributions. The statement was only that if we get something like this on short notice, it is hard to guarantee that we can release a patch the same day.

**Comment by JeffAshton** [comment]
> @JeffAshton, my opinion is the same as Louis's. No one said you should not submit PRs; we've always appreciated your contributions. The statement was only that if we get something like this on short notice, it is hard to guarantee that we can release a patch the same day.

It was just a request.  You can always say no Eli.  We have history.  :)

**Comment by eli-darkly** [comment]
@JeffAshton, we understand that it is a request, and we have been working on it. If you reread Louis's comment, which again is the same one I would have made, I think you'll see that there's no need to take it as "infuriating"; he was simply responding to the "Can I ask you to cut a release today" part to clarify that _if possible_ it's always best for us to have slightly more notice— a message that's directed to all customers equally, since this is a public conversation. That, too, was just a request on our part, and in cases where it's not possible, fine.

**Comment by eli-darkly** [comment]
Version 6.3.2 has been released with this fix.

**Comment by eli-darkly** [comment]
Also, since there was a period of time when we had a smaller SDK team and I was the only maintainer of this project, I should clarify that that hasn't been the case for some time now. The development process and release scheduling happens collaboratively, and we'll see issues/PRs regardless of whether I am @-mentioned.

**Pull request closed without merging**